src/typeorm-aurora-data-api-driver.ts

```diff
@@ -1,95 +1,113 @@
 // @ts-ignore
 import createDataApiClient from 'data-api-client'
 
-const quoteCharacters = ['\'', '"']
-
-export default class DataApiDriver {
-  public static transformQueryAndParameters(query: string, parameters?: any[]): any {
-    let numberOfParametersInQueryString = 0
-
-    let newQueryString = ''
-    let currentQuote = null
-
-    for (let i = 0; i < query.length; i += 1) {
-      const currentCharacter = query[i]
-      const currentCharacterEscaped = i !== 0 && query[i - 1] === '\\'
-
-      if (currentCharacter === '?' && !currentQuote) {
+const transformQuery = (query: string, parameters?: any[]): [string, number] => {
+  const quoteCharacters = ['\'', '"']
+
+  let newQueryString = ''
+  let currentQuote = null
+  let numberOfParametersInQueryString = 0
+  let extendedParameterCount = 0
+
+  for (let i = 0; i < query.length; i += 1) {
+    const currentCharacter = query[i]
+    const currentCharacterEscaped = i !== 0 && query[i - 1] === '\\'
+
+    if (currentCharacter === '?' && !currentQuote) {
+      const parameter = parameters![numberOfParametersInQueryString]
+      if (Array.isArray(parameter)) {
+        parameter.forEach((_, index) => {
+          const isLastParameter = index === parameter.length - 1
+          const paramName = `:e_param_${extendedParameterCount}${isLastParameter ? '' : ', '}`
+          extendedParameterCount += 1
+          newQueryString += paramName
+        })
+      } else {
         const paramName = `:param_${numberOfParametersInQueryString}`
-
         numberOfParametersInQueryString += 1
         newQueryString += paramName
-      } else if (quoteCharacters.includes(currentCharacter) && !currentCharacterEscaped) {
-        if (!currentQuote) {
-          currentQuote = currentCharacter
-        } else if (currentQuote === currentCharacter) {
-          currentQuote = null
-        }
-
-        newQueryString += currentCharacter
-      } else {
-        newQueryString += currentCharacter
       }
+
+      continue
     }
 
-    if (
-      parameters &&
-      parameters.length > 0 &&
-      parameters.length % numberOfParametersInQueryString !== 0
-    ) {
-      throw new Error(`Number of parameters mismatch, got ${numberOfParametersInQueryString} in query string \
-            and ${parameters.length} in input`)
+    if (quoteCharacters.includes(currentCharacter) && !currentCharacterEscaped) {
+      if (!currentQuote) {
+        currentQuote = currentCharacter
+      } else if (currentQuote === currentCharacter) {
+        currentQuote = null
+      }
     }
 
-    const transformedParameters: any[] = []
+    newQueryString += currentCharacter
+  }
+
+  return [newQueryString, numberOfParametersInQueryString]
+}
+
+const transformParameters = (numberOfParametersInQueryString: number, parameters?: any[]) => {
+  if (
+    parameters &&
+    parameters.length > 0 &&
+    parameters.length % numberOfParametersInQueryString !== 0
```

Comment:
We need to check against expanded parameters here, because the number of parameters in a query string may already be bigger than the one in an array.
I would suggest to make that check after we do the transformation

Reply:
Please see my above comment. This logic should be exactly the same as before. I didn't want to change anything that was obviously working fine. I do not think the expanded parameters need to be given consideration here.  Please correct me if I am mistaken.